text/0097-universal-wallet.md

```diff
@@ -0,0 +1,56 @@
+- **TEP**: [0](https://github.com/ton-blockchain/TEPs/pull/0) *(don't change)*
+- **title**: Universal Wallet
+- **status**: Draft
+- **type**: Core
+- **authors**: [Steve Korshakov](https://github.com/ex3ndr) *(replace)*
+- **created**: 06.10.2022
+- **replaces**: [TEP-0](https://github.com/ton-blockchain/TEPs/blob/master/0000-template.md)
+- **replaced by**: -
+
+# Summary
+
+This TEP introduces universal wallet contract that can be upgraded by validator voting.
+
+# Motivation
+
+Currently TON have multiple generations of wallet contracts and it was very painful to upgrade users from old to a newer one. Currently even after 4 generations and dozen of revisions it is still incomplete and from time to time requires upgrades. Upgrading currently involves change of address and requires to move all your assets to a new address and could cause a lot of mess and non-managebale for average user.
+
+# Guide
+
+This universal wallet is implemented as FIFT code that fetches code from config param #74 and executes it.
+
+# Specification
+
+Code for a universal wallet is placed as critical config #74 and could be upgraded by voting by validators.
+
+The wallet itself have code (Fift):
+```
+PUSHINT 74
+CONFIGPARAM
+BLESS
+JMPX
+```
+
+And inital data is `seq:int64 public_key:int256 wallet_id:int64`.
```

Comment:
What if in new version of the wallet there will be more variables in data? How will it upgrade then?

Reply:
Code must support the absence of any added field. We are doing this in our nominators pool and it worked well.